### ISSUE TITLE
vm with multiple disks

### ISSUE THREAD

**Issue by leagueofsoups** [OpenNebula/terraform-provider-opennebula#264]
Hello. I attach the file below

- every `terraform apply` will create new vd* disk
- when uncomment `target           = "vdb"` on disk description, and remove `dev_prefix   = "vd"` from opennebula_image.img_data resource, second `terraform apply` return error.

It looks like there is an idempotency problem
The problem is reproduced on OpenNebula 6.2.0 ce

```
terraform {required_providers {
    opennebula = {
      source = "opennebula/opennebula"
      version = "0.4.3"
    }
  }
}

provider "opennebula" {
  endpoint      = "FIXIT"
  flow_endpoint = "FIXIT"
  username      = "FIXIT"
  password      = "FIXIT"
}

resource "opennebula_image" "img" {
    name        = "ubuntu_base"
    datastore_id = 1
    persistent   = false
    lock         = "MANAGE"
    path         = "https://marketplace.opennebula.io/appliance/695f1a36-b970-4ccf-ace3-0863dcc86d2a/download/0"
    timeout      = 300
}

resource "opennebula_image" "img_clon" {
    clone_from_image = opennebula_image.img.id
    datastore_id     = 1
    name             = "os"
}

resource "opennebula_image" "img_data" {
    datastore_id     = 1
    name             = "data"
    lock         = "MANAGE"
    type         = "DATABLOCK"
    timeout      = 300
    size         = 3000
    dev_prefix   = "vd"

}

resource "opennebula_virtual_machine" "k8s-domain-ubuntu" {

  name        = "vm"
  
  cpu         = 1
  vcpu        = 1
  memory      = 1000

  context = {
    NETWORK        = "YES"
    SET_HOSTNAME   = "$NAME",
    HOSTNAME       = "$NAME"
    PASSWORD       = "root"
  }

  graphics {
    type   = "VNC"
    listen = "0.0.0.0"
  }

  os {
    arch = "x86_64"
    boot = "disk0"
  }


  disk {
    image_id        = opennebula_image.img_clon.id
    target           = "vda"
    driver           = "qcow2"
    size             = 3000
  }

  disk {
    image_id         = opennebula_image.img_data.id
    #target           = "vdb"
    driver           = "qcow2"
  }

  timeout = 15
}
```

**Comment by frousselet** [comment]
Hello @leagueofsoups,

Thanks for submitting this issue.

Could you provide a `terraform plan` output please?
Also, when you said _'every `terraform apply` will create new vd* disk'_, did you mean destroy/recreate in place or duplicate?

_'when uncomment target = "vdb" on disk description, and remove dev_prefix = "vd" from opennebula_image.img_data resource, second terraform apply return error.'_

Could you show us the errors and the [debug](https://www.terraform.io/internals/debugging) output of `terraform apply`, with `TF_LOG=DEBUG` please?

Thanks.

**Comment by leagueofsoups** [comment]
Hello @frousselet,

Thanks for your reply. 

When I use manifest from my first post:
```
terraform apply -auto-approve
terraform plan
```

```
opennebula_image.img: Refreshing state... [id=465]
opennebula_image.img_data: Refreshing state... [id=464]
opennebula_image.img_clon: Refreshing state... [id=466]
opennebula_virtual_machine.k8s-domain-ubuntu: Refreshing state... [id=242]

Terraform used the selected providers to generate the following execution plan. Resource
actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opennebula_virtual_machine.k8s-domain-ubuntu will be updated in-place
  ~ resource "opennebula_virtual_machine" "k8s-domain-ubuntu" {
        id            = "242"
        name          = "vm"
        # (16 unchanged attributes hidden)

      ~ disk {
          - volatile_format = "qcow2" -> null
            # (8 unchanged attributes hidden)
        }
      + disk {
          + driver   = "qcow2"
          + image_id = 464
        }


        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```
Terraform will create new disk every time, when `terraform apply`
For example:
```
terraform destroy
terraform apply -auto-approve
terraform apply -auto-approve
terraform apply -auto-approve
```
result:
![image](https://user-images.githubusercontent.com/41836635/167627089-29179aad-00f5-430e-8c08-eacdbb5f0a67.png)

When uncomment `target = "vdb"` on disk description, and remove `dev_prefix = "vd"` from opennebula_image.img_data resource, second terraform apply return error.

```
terraform destroy
terraform apply -auto-approve
terraform plan
TF_LOG=debug terraform apply -auto-approve
```
plan:
```
opennebula_image.img: Refreshing state... [id=471]
opennebula_image.img_data: Refreshing state... [id=470]
opennebula_image.img_clon: Refreshing state... [id=472]
opennebula_virtual_machine.k8s-domain-ubuntu: Refreshing state... [id=244]

Terraform used the selected providers to generate the following execution plan. Resource
actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opennebula_virtual_machine.k8s-domain-ubuntu will be updated in-place
  ~ resource "opennebula_virtual_machine" "k8s-domain-ubuntu" {
        id            = "244"
        name          = "vm"
        # (16 unchanged attributes hidden)

      ~ disk {
          - volatile_format = "qcow2" -> null
            # (8 unchanged attributes hidden)
        }
      + disk {
          + driver   = "qcow2"
          + image_id = 470
          + target   = "vdb"
        }


        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```
error:
```
2022-05-10T15:37:46.899+0300 [INFO]  Terraform version: 1.1.9
2022-05-10T15:37:46.899+0300 [INFO]  Go runtime version: go1.17.2
2022-05-10T15:37:46.899+0300 [INFO]  CLI args: []string{"terraform", "apply", "-auto-approve"}
2022-05-10T15:37:46.899+0300 [DEBUG] Attempting to open CLI config file: /Users/sergejsmirnov/.terraformrc
2022-05-10T15:37:46.899+0300 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2022-05-10T15:37:46.899+0300 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2022-05-10T15:37:46.899+0300 [DEBUG] ignoring non-existing provider search directory /Users/sergejsmirnov/.terraform.d/plugins
2022-05-10T15:37:46.899+0300 [DEBUG] ignoring non-existing provider search directory /Users/sergejsmirnov/Library/Application Support/io.terraform/plugins
2022-05-10T15:37:46.899+0300 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2022-05-10T15:37:46.899+0300 [INFO]  CLI command args: []string{"apply", "-auto-approve"}
2022-05-10T15:37:46.901+0300 [DEBUG] New state was assigned lineage "9542ee4f-d540-4851-4530-14faa63003ae"
2022-05-10T15:37:46.959+0300 [DEBUG] checking for provisioner in "."
2022-05-10T15:37:46.964+0300 [DEBUG] checking for provisioner in "/usr/local/bin"
2022-05-10T15:37:46.965+0300 [INFO]  backend/local: starting Apply operation
2022-05-10T15:37:46.968+0300 [DEBUG] created provider logger: level=debug
2022-05-10T15:37:46.968+0300 [INFO]  provider: configuring client automatic mTLS
2022-05-10T15:37:46.978+0300 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 args=[.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3]
2022-05-10T15:37:46.981+0300 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 pid=65146
2022-05-10T15:37:46.981+0300 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3
2022-05-10T15:37:46.993+0300 [INFO]  provider.terraform-provider-opennebula_v0.4.3: configuring server automatic mTLS: timestamp=2022-05-10T15:37:46.993+0300
2022-05-10T15:37:47.003+0300 [DEBUG] provider: using plugin: version=5
2022-05-10T15:37:47.003+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: plugin address: address=/tmp/plugin2976187897 network=unix timestamp=2022-05-10T15:37:47.002+0300
2022-05-10T15:37:47.017+0300 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unimplemented desc = unknown service plugin.GRPCStdio"
2022-05-10T15:37:47.019+0300 [DEBUG] No provider meta schema returned
2022-05-10T15:37:47.021+0300 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 pid=65146
2022-05-10T15:37:47.021+0300 [DEBUG] provider: plugin exited
2022-05-10T15:37:47.021+0300 [DEBUG] Building and walking validate graph
2022-05-10T15:37:47.022+0300 [DEBUG] ProviderTransformer: "opennebula_image.img_data" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.022+0300 [DEBUG] ProviderTransformer: "opennebula_virtual_machine.k8s-domain-ubuntu" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.022+0300 [DEBUG] ProviderTransformer: "opennebula_image.img" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.022+0300 [DEBUG] ProviderTransformer: "opennebula_image.img_clon" (*terraform.NodeValidatableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.022+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img_clon" references: [opennebula_image.img]
2022-05-10T15:37:47.022+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img_data" references: []
2022-05-10T15:37:47.022+0300 [DEBUG] ReferenceTransformer: "opennebula_virtual_machine.k8s-domain-ubuntu" references: [opennebula_image.img_clon opennebula_image.img_data]
2022-05-10T15:37:47.022+0300 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/opennebula/opennebula\"]" references: []
2022-05-10T15:37:47.022+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img" references: []
2022-05-10T15:37:47.022+0300 [DEBUG] Starting graph walk: walkValidate
2022-05-10T15:37:47.023+0300 [DEBUG] created provider logger: level=debug
2022-05-10T15:37:47.023+0300 [INFO]  provider: configuring client automatic mTLS
2022-05-10T15:37:47.032+0300 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 args=[.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3]
2022-05-10T15:37:47.034+0300 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 pid=65147
2022-05-10T15:37:47.034+0300 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3
2022-05-10T15:37:47.046+0300 [INFO]  provider.terraform-provider-opennebula_v0.4.3: configuring server automatic mTLS: timestamp=2022-05-10T15:37:47.045+0300
2022-05-10T15:37:47.055+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: plugin address: address=/tmp/plugin3054652888 network=unix timestamp=2022-05-10T15:37:47.055+0300
2022-05-10T15:37:47.055+0300 [DEBUG] provider: using plugin: version=5
2022-05-10T15:37:47.070+0300 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unimplemented desc = unknown service plugin.GRPCStdio"
2022-05-10T15:37:47.071+0300 [DEBUG] No provider meta schema returned
2022-05-10T15:37:47.081+0300 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 pid=65147
2022-05-10T15:37:47.081+0300 [DEBUG] provider: plugin exited
2022-05-10T15:37:47.081+0300 [INFO]  backend/local: apply calling Plan
2022-05-10T15:37:47.081+0300 [DEBUG] Building and walking plan graph for NormalMode
2022-05-10T15:37:47.081+0300 [DEBUG] ProviderTransformer: "opennebula_image.img (expand)" (*terraform.nodeExpandPlannableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.081+0300 [DEBUG] ProviderTransformer: "opennebula_image.img_clon (expand)" (*terraform.nodeExpandPlannableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.081+0300 [DEBUG] ProviderTransformer: "opennebula_image.img_data (expand)" (*terraform.nodeExpandPlannableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.081+0300 [DEBUG] ProviderTransformer: "opennebula_virtual_machine.k8s-domain-ubuntu (expand)" (*terraform.nodeExpandPlannableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.081+0300 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/opennebula/opennebula\"]" references: []
2022-05-10T15:37:47.081+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img (expand)" references: []
2022-05-10T15:37:47.082+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img_clon (expand)" references: [opennebula_image.img (expand)]
2022-05-10T15:37:47.082+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img_data (expand)" references: []
2022-05-10T15:37:47.082+0300 [DEBUG] ReferenceTransformer: "opennebula_virtual_machine.k8s-domain-ubuntu (expand)" references: [opennebula_image.img_clon (expand) opennebula_image.img_data (expand)]
2022-05-10T15:37:47.082+0300 [DEBUG] Starting graph walk: walkPlan
2022-05-10T15:37:47.082+0300 [DEBUG] created provider logger: level=debug
2022-05-10T15:37:47.082+0300 [INFO]  provider: configuring client automatic mTLS
2022-05-10T15:37:47.091+0300 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 args=[.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3]
2022-05-10T15:37:47.094+0300 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 pid=65148
2022-05-10T15:37:47.094+0300 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3
2022-05-10T15:37:47.106+0300 [INFO]  provider.terraform-provider-opennebula_v0.4.3: configuring server automatic mTLS: timestamp=2022-05-10T15:37:47.106+0300
2022-05-10T15:37:47.116+0300 [DEBUG] provider: using plugin: version=5
2022-05-10T15:37:47.116+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: plugin address: address=/tmp/plugin62737892 network=unix timestamp=2022-05-10T15:37:47.116+0300
2022-05-10T15:37:47.130+0300 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unimplemented desc = unknown service plugin.GRPCStdio"
2022-05-10T15:37:47.131+0300 [DEBUG] No provider meta schema returned
2022-05-10T15:37:47.135+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img_data" references: []
2022-05-10T15:37:47.135+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img" references: []
2022-05-10T15:37:47.151+0300 [WARN]  Provider "registry.terraform.io/opennebula/opennebula" produced an invalid plan for opennebula_image.img_data, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .persistent: planned value cty.False for a non-computed attribute
2022-05-10T15:37:47.152+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img_clon" references: []
2022-05-10T15:37:47.164+0300 [WARN]  Provider "registry.terraform.io/opennebula/opennebula" produced an invalid plan for opennebula_image.img_clon, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .persistent: planned value cty.False for a non-computed attribute
      - .timeout: planned value cty.NumberIntVal(10) for a non-computed attribute
2022-05-10T15:37:47.164+0300 [DEBUG] ReferenceTransformer: "opennebula_virtual_machine.k8s-domain-ubuntu" references: []
2022-05-10T15:37:47.175+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:37:47 [WARN] Configuration for disk ID: 1 not found.
2022-05-10T15:37:47.181+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:37:47 [WARN] Configuration for disk ID: 1 not found.
2022-05-10T15:37:47.190+0300 [WARN]  Provider "registry.terraform.io/opennebula/opennebula" produced an invalid plan for opennebula_virtual_machine.k8s-domain-ubuntu, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .template_id: planned value cty.NumberIntVal(-1) for a non-computed attribute
      - .pending: planned value cty.False for a non-computed attribute
      - .disk[0].volatile_type: planned value cty.StringVal("") for a non-computed attribute
      - .graphics[0].keymap: planned value cty.StringVal("en-us") for a non-computed attribute
2022-05-10T15:37:47.193+0300 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 pid=65148
2022-05-10T15:37:47.193+0300 [DEBUG] provider: plugin exited
2022-05-10T15:37:47.197+0300 [INFO]  backend/local: apply calling Apply
2022-05-10T15:37:47.197+0300 [DEBUG] Building and walking apply graph for NormalMode plan
2022-05-10T15:37:47.198+0300 [DEBUG] ProviderTransformer: "opennebula_virtual_machine.k8s-domain-ubuntu (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.198+0300 [DEBUG] ProviderTransformer: "opennebula_image.img (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.198+0300 [DEBUG] ProviderTransformer: "opennebula_virtual_machine.k8s-domain-ubuntu" (*terraform.NodeApplyableResourceInstance) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.198+0300 [DEBUG] ProviderTransformer: "opennebula_image.img_clon (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.198+0300 [DEBUG] ProviderTransformer: "opennebula_image.img_data (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/opennebula/opennebula"]
2022-05-10T15:37:47.198+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img_clon (expand)" references: []
2022-05-10T15:37:47.198+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img_data (expand)" references: []
2022-05-10T15:37:47.198+0300 [DEBUG] ReferenceTransformer: "opennebula_virtual_machine.k8s-domain-ubuntu (expand)" references: []
2022-05-10T15:37:47.198+0300 [DEBUG] ReferenceTransformer: "opennebula_image.img (expand)" references: []
2022-05-10T15:37:47.198+0300 [DEBUG] ReferenceTransformer: "opennebula_virtual_machine.k8s-domain-ubuntu" references: [opennebula_image.img_clon (expand) opennebula_image.img_data (expand)]
2022-05-10T15:37:47.198+0300 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/opennebula/opennebula\"]" references: []
2022-05-10T15:37:47.198+0300 [DEBUG] pruneUnusedNodes: opennebula_image.img (expand) is no longer needed, removing
2022-05-10T15:37:47.198+0300 [DEBUG] Starting graph walk: walkApply
2022-05-10T15:37:47.199+0300 [DEBUG] created provider logger: level=debug
2022-05-10T15:37:47.199+0300 [INFO]  provider: configuring client automatic mTLS
2022-05-10T15:37:47.208+0300 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 args=[.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3]
2022-05-10T15:37:47.211+0300 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 pid=65149
2022-05-10T15:37:47.212+0300 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3
2022-05-10T15:37:47.224+0300 [INFO]  provider.terraform-provider-opennebula_v0.4.3: configuring server automatic mTLS: timestamp=2022-05-10T15:37:47.223+0300
2022-05-10T15:37:47.236+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: plugin address: address=/tmp/plugin2685694070 network=unix timestamp=2022-05-10T15:37:47.235+0300
2022-05-10T15:37:47.236+0300 [DEBUG] provider: using plugin: version=5
2022-05-10T15:37:47.249+0300 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unimplemented desc = unknown service plugin.GRPCStdio"
2022-05-10T15:37:47.250+0300 [DEBUG] No provider meta schema returned
2022-05-10T15:37:47.268+0300 [WARN]  Provider "registry.terraform.io/opennebula/opennebula" produced an invalid plan for opennebula_virtual_machine.k8s-domain-ubuntu, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .pending: planned value cty.False for a non-computed attribute
      - .template_id: planned value cty.NumberIntVal(-1) for a non-computed attribute
      - .graphics[0].keymap: planned value cty.StringVal("en-us") for a non-computed attribute
      - .disk[0].volatile_type: planned value cty.StringVal("") for a non-computed attribute
2022-05-10T15:37:47.270+0300 [INFO]  Starting apply for opennebula_virtual_machine.k8s-domain-ubuntu
2022-05-10T15:37:47.271+0300 [DEBUG] opennebula_virtual_machine.k8s-domain-ubuntu: applying the planned Update change
2022-05-10T15:37:47.281+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:37:47 [INFO] Update disk configuration
2022-05-10T15:37:47.281+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:37:47 [DEBUG] Detach disk 0
2022-05-10T15:37:47.529+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:37:47 [DEBUG] Waiting for state to become: [RUNNING POWEROFF]
2022-05-10T15:37:57.529+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:37:57 Refreshing VM state...
2022-05-10T15:37:57.536+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:37:57 VM (ID:244, name:vm) is currently in state ACTIVE and in LCM state RUNNING
2022-05-10T15:37:57.543+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:37:57 [DEBUG] Attach disk to virtual machine (ID:244)
2022-05-10T15:37:57.946+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:37:57 [DEBUG] Waiting for state to become: [RUNNING POWEROFF]
2022-05-10T15:38:07.946+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:38:07 Refreshing VM state...
2022-05-10T15:38:07.957+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:38:07 VM (ID:244, name:vm) is currently in state ACTIVE and in LCM state RUNNING
2022-05-10T15:38:07.967+0300 [DEBUG] provider.terraform-provider-opennebula_v0.4.3: 2022/05/10 15:38:07 [DEBUG] Attach disk to virtual machine (ID:244)
2022-05-10T15:38:08.222+0300 [ERROR] vertex "opennebula_virtual_machine.k8s-domain-ubuntu" error: vm disk attach: can't attach image to virtual machine (ID:244): OpenNebula error [ACTION]: [one.vm.attach] Target vdb is already in use.
╷
│ Error: vm disk attach: can't attach image to virtual machine (ID:244): OpenNebula error [ACTION]: [one.vm.attach] Target vdb is already in use.
│ 
│ 
│   with opennebula_virtual_machine.k8s-domain-ubuntu,
│   on tf-debug.tf line 43, in resource "opennebula_virtual_machine" "k8s-domain-ubuntu":
│   43: resource "opennebula_virtual_machine" "k8s-domain-ubuntu" {
│ 
╵
2022-05-10T15:38:08.251+0300 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/opennebula/opennebula/0.4.3/darwin_amd64/terraform-provider-opennebula_v0.4.3 pid=65149
2022-05-10T15:38:08.251+0300 [DEBUG] provider: plugin exited
```

**Comment by treywelsh** [comment]
Thank you for your detailed report it really helps, and sorry for this obvious error

Here is an early feedback, I didn't try all cases you mention:
I'm able to reproduce an unexpected diff on `volatile_format` attribute with my current env:
TF: 0.15.5
Provider: 0.4.3
ONE: 6.2.0

After the apply, at plan I get:
```
opennebula_image.img: Refreshing state... [id=1]
opennebula_image.img_data: Refreshing state... [id=7]
opennebula_image.img_clon: Refreshing state... [id=8]
opennebula_virtual_machine.k8s-domain-ubuntu: Refreshing state... [id=6]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opennebula_virtual_machine.k8s-domain-ubuntu will be updated in-place
  ~ resource "opennebula_virtual_machine" "k8s-domain-ubuntu" {
        id            = "6"
        name          = "vm"
        # (16 unchanged attributes hidden)

      ~ disk {
          - volatile_format = "qcow2" -> null
            # (8 unchanged attributes hidden)
        }
      ~ disk {
          - volatile_format = "qcow2" -> null
            # (7 unchanged attributes hidden)
        }


        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

It tried again but this time I built the provider on top of master branch (we're currently working on 0.5.0 release of the provider), it seems that there is no changes anymore, I suspect this problem to be fixed by PR
https://github.com/OpenNebula/terraform-provider-opennebula/pull/260 merged some days ago.

If you are able to build the provider and test it in an harmless environement, feel free to test (you could only cherry-pick the commit of the PR mentionned above on top of a branch just above the v0.4.3 tag) and give us a feedback.

This will need some more tests/investigations to be sure

**Comment by treywelsh** [comment]
To give a bit more informations on my current understanding of the problem:
Before merging #260, the `volatile_format` key was read everytime.
When I introduced volatile disk (for context, they were introduced by PR #155) I didn't expected the attribute `volatile_format` to have a value even for non volatile disks. (In OpenNebula virtual machine template, for a disk, the attribute name is `FORMAT`, in fact `volatile_format` is the name of the attribute in the provider schema)

The PR #260 introduce the `computed_volatile_format` that receive the value everytime (for volatile and non volatile disks), and that fill the `volatile_format` only when the user has configured this value. This allow to see if the applied value match the user configuration.

**Comment by leagueofsoups** [comment]
Hello @treywelsh 
I can't reproduce problem after rebuild provider from current master branch. 
The description of the problem disk looks like this:
```
  disk {
    target          = "vdb"
    driver          = "qcow2"
    volatile_type   = "fs"
    volatile_format = "qcow2"
    size            = 3000
  }
```
It seems that the issue can be closed, because it is not reproduced in the 0.5.0 version of the provider.
Thank you very much for your help!
